core/query/evalNestedQuery.ml

```diff
@@ -487,9 +489,9 @@ struct
     function
       | Constant c    -> Constant c
       | Primitive p   -> Primitive p
-      | Apply ("Empty", [e]) -> Apply ("Empty", [flatten_inner_query e])
-      | Apply ("length", [e]) -> Apply ("length", [flatten_inner_query e])
-      | Apply (f, es) -> Apply (f, List.map flatten_inner es)
+      | Apply (Primitive "Empty", [e]) -> Apply (Primitive "Empty", [flatten_inner_query e])
+      | Apply (Primitive "length", [e]) -> Apply (Primitive "length", [flatten_inner_query e])
+      | Apply (Primitive f, es) -> Apply (Primitive f, List.map flatten_inner es)
```

Comment:
Assuming we operate on normalized queries, the first argument of `Apply` _must_ be a Primitive (if it's not, then there's a problem somewhere else). On the other hand, I don't really see why we need the extra cases for "Empty" and "length", though it could be a (minimal) optimization.

Reply:
OK.  It looks like here the first two, specialized, cases are now formally redundant, except perhaps that they check that empty and length have exactly one argument (something that should already have been checked earlier).

Reply:
On second thought, this and lines 385-389 should stay as is, since the Empty and length primitives are treated differently than other primitives.  (using *_inner_query instead of *_inner).